doc/dev/mountpoints.md

```diff
@@ -0,0 +1,179 @@
+
+Every mountpoint consists of exactly three parts:
+
+1. A list of plugins required for the mountpoint.
+2. A single _backend plugin_ chosen from the list of plugins.
+3. The _mountpoint definition_ specific to the backend plugin.
+
+The list of plugins is defined as the array `system:/elektra/mountpoints/<mountpoint>/plugins/#`.
+Specifically, `system:/elektra/mountpoints/<mountpoint>/plugins/#/name` contains the name of the plugin and `system:/elektra/mountpoints/<mountpoint>/plugins/#/config` the configuration keyset passed to the plugin.
+
+The single _backend plugin_ is the only one that the `libelektra-kdb` library communicates with.
+It is responsible for calling other plugins when needed.
+A _backend contract_ exists between this plugin and `libelektra-kdb`.
+The backend plugin alone is responsible for upholding this contract.
+
+The backend plugin must be one of the ones defined in `system:/elektra/plugins/#`.
+The tell `libelekta-kdb` which of the plugins to use as the _backend plugin_, you set `system:/elektra/mountpoints/<mountpoint>/backend` to its array index.
+
+All other keys below `system:/elektra/mountpoints/<mountpoint>` are the _mountpoint definition_ for the backend plugin.
+
+> **Note**: This _mountpoint definition_ is separate from the normal configuration keyset passed to a plugin.
+> The backend plugin may still have such a keyset below its `system:/elektra/mountpoints/<mountpoint>/plugins/#/config` key.
+> The difference between these two keyset is that the configuration keyset should be used to define the general operation of the plugin, while the mountpoint definition should specify a mountpoint.
+>
+> For example, a backend plugin could support a logging mode, in which it produces a log entry every time it calls another plugin.
+> This should be configured in the configuration keyset not in the mountpoint definition, since it doesn't change how the mountpoint works.
+
+To illustrate this structure, here is an example configuration:
+
+```
+# List of plugins with their config
+system:/elektra/mountpoints/\/hosts/plugins/#0/name (="resolver_fm_hpu_b")
+system:/elektra/mountpoints/\/hosts/plugins/#1/name (="glob")
+system:/elektra/mountpoints/\/hosts/plugins/#1/config/set/#0
+system:/elektra/mountpoints/\/hosts/plugins/#1/config/set/#1
+system:/elektra/mountpoints/\/hosts/plugins/#1/config/set/#2
+system:/elektra/mountpoints/\/hosts/plugins/#1/config/set/#3
+system:/elektra/mountpoints/\/hosts/plugins/#1/config/set/#4/flags
+system:/elektra/mountpoints/\/hosts/plugins/#2/name (="hosts")
+system:/elektra/mountpoints/\/hosts/plugins/#3/name (="sync")
+system:/elektra/mountpoints/\/hosts/plugins/#4/name (="error")
+system:/elektra/mountpoints/\/hosts/plugins/#5/name (="network")
+system:/elektra/mountpoints/\/hosts/plugins/#6/name (="backend")
+
+# Define backend plugin
+system:/elektra/mountpoints/\/hosts/backend (="#6")
+
+# Configuration for backend plugin
+system:/elektra/mountpoints/\/hosts/path (="myhosts")
```

Comment:
Please put to config of backend, then we only have 3 KeySets: plugin, config, position

Having backend and plugin config is very useful.

Reply:
I do not understand at all want you want to say here.

Reply:
Mountpoints (backends) used to have two different kinds of config:

- plugin-specific configs (copied to user:)
- backend-specific configs (copied to system:)

I did not see the place in your proposal where you allow backend-specific configs.

Furthermore, I would like that the "path" is in this backed-specific config. I would prefer if there are not too many top-level keys below `/elektra/mountpoints/<backend>` to ease future extensions (which would be impossible if we allow any key below there).

Reply:
What exactly is "backend specific" config used for? Why do we need it? Where does it come from?

---

> I would prefer if there are not too many top-level keys below `/elektra/mountpoints/<backend>` to ease future extensions (which would be impossible if we allow any key below there).

I'm not sure why we would need a future extension. By design `libelektra-kdb` loads all plugins and then only talks to the designated backend plugin. There isn't really anything that could be added here.

If you just want to be on the safe side, we could keep `/elektra/mountpoints/<backend>/plugins` and `/elektra/mountpoints/<backend>/backend` and move the "mountpoint definition" (i.e. the other keys below `/elektra/mountpoints/<backend>`) to `/elektra/mountpoints/<backend>/definition`.

Reply:
> What exactly is "backend specific" config used for? Why do we need it? Where does it come from?

There are two sources:

1. The `-c` switch of `kdb mount` so that users can pass the same config to all plugins
2. Plugins can create it via `config/needs`. It allows plugins to configure other plugins, e.g. the hosts plugin configures the glob+network plugin that way.

Reply:
IMO these are two different features that should be supported separately.

For the first part, we should probably use e.g. `/elektra/mountpoints/<backend>/config` for config that is provided to all plugins. For the second part I'm not really sure what to do. AFAIK `config/needs` is processed by the tools and not by `libelektra-kdb`. Maybe we should just change that and read this on the fly (i.e. not store it `/elektra/mountpoints/<backend>` at all). Then you also wouldn't have to remount a mountpoint, if a plugin changes it's `config/needs`.

Reply:
Discussion result: `/elektra/mountpoints/<backend>/config` will be added. `config/needs` will be read on-the-fly.

Additional results: The difference between "config" and "definition" is: "config" is processed during `kdbOpen`, "definition" only during the first `kdbGet`. Therefore, "config" is always processed for all plugins in all mountpoints, but "definition" only for those mountpoints that are actually accessed. This also means that `elektra<Plugin>Open` should _not _ create any expensive resources (e.g. DB connections), this should happen lazily during the first `elektra<Plugin>Get` in the `init` phase.